azure-core/azure/core/pipeline/policies/universal.py

```diff
@@ -0,0 +1,310 @@
+import contextlib
+import json
+import logging
+import os
+import platform
+import threading
+from typing import TYPE_CHECKING, cast, List, Callable, Iterator, Any, Union, Dict, Optional  # pylint: disable=unused-import
+import xml.etree.ElementTree as ET
+import warnings
+import types
+import re
+
+from azure.core import __version__  as azcore_version
+from .base import HTTPPolicy, SansIOHTTPPolicy
+from urllib3 import Retry  # Needs requests 2.16 at least to be safe
+
+from azure.core.exceptions import (
+    DecodeError,
+    raise_with_traceback
+)
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class HeadersPolicy(SansIOHTTPPolicy):
+    """A simple policy that sends the given headers
+    with the request.
+
+    This overwrite any headers already defined in the request.
+    """
+    def __init__(self, headers=None, **kwargs):
+        # type: (Mapping[str, str]) -> None
+        self.headers = headers or {}
+
+    def on_request(self, request, **kwargs):
+        # type: (PipelineRequest, Any) -> None
+        additional_headers = kwargs.pop('headers', {})
+        http_request = request.http_request
+        http_request.headers.update(self.headers)
+        http_request.headers.update(additional_headers)
+
+
+class UserAgentPolicy(SansIOHTTPPolicy):
+    _USERAGENT = "User-Agent"
+    _ENV_ADDITIONAL_USER_AGENT = 'AZURE_HTTP_USER_AGENT'
+
+    def __init__(self, base_user_agent=None, **kwargs):
+        # type: (Optional[str], bool) -> None
+        self.overwrite = kwargs.pop('user_agent_overwrite', False)
+        self.use_env = kwargs.pop('user_agent_use_env', True)
+
+        if base_user_agent is None:
+            self._user_agent = "python/{} ({}) azure-core/{}".format(
+                platform.python_version(),
+                platform.platform(),
+                azcore_version
+            )
+        else:
+            self._user_agent = base_user_agent
+
+    @property
+    def user_agent(self):
+        # type: () -> str
+        """The current user agent value."""
+        if self.use_env:
+            add_user_agent_header = os.environ.get(self._ENV_ADDITIONAL_USER_AGENT, None)
+            if add_user_agent_header is not None:
+                return "{} {}".format(self._user_agent, add_user_agent_header)
+        return self._user_agent
+
+    def add_user_agent(self, value):
+        # type: (str) -> None
+        """Add value to current user agent with a space.
+
+        :param str value: value to add to user agent.
+        """
+        self._user_agent = "{} {}".format(self._user_agent, value)
+
+    def on_request(self, request, **kwargs):
+        # type: (PipelineRequest, Any) -> None
+        http_request = request.http_request
+        if 'user_agent' in kwargs:
+            user_agent = kwargs.pop('user_agent')
+            if kwargs.pop('user_agent_overwrite', self.overwrite):
+                http_request.headers[self._USERAGENT] = user_agent
```

Comment:
Header names are case insensitive. Which means that one could have both a `user-agent` and `User-Agent` header, and we would only override one of them. Or add a duplicate key. In practical terms, I don't know if this matters, though...

Reply:
Never mind - we are using a case insensitive dict.